### PR TITLE
Allow user to specify the Name tag value for AWS tags

### LINKED ISSUE #3989: User-provided `Name` tag on subnets get overwritten
/kind bug

**What steps did you take and what happened:**

Specify a subnet on an `AWSCluster` CR that has a `Name` tag defined. E.g. 

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AWSCluster
metadata:
  name: example
spec:
  network:
    subnets:
    - availabilityZone: eu-north-1a
      cidrBlock: 10.0.0.0/24
      isPublic: false
      tags:
        Name: example-custom-subnet-name
```

This `Name` tag will be instead replaced with one built in code and instead become `Name: example-subnet-private-eu-north-1a`

**What did you expect to happen:**

If a `Name` tag is provided by the user it is what should be used when creating / updating the subnet. If not provided CAPA should default to the current generated format.

**Anything else you would like to add:**

Block of code building the Name value: https://github.com/giantswarm/cluster-api-provider-aws/blob/70373384dc117ea387d28a8548d49ecc58e7cca0/pkg/cloud/services/network/subnets.go#L426-L441

Block of code overwriting any user-provided Name tag: https://github.com/giantswarm/cluster-api-provider-aws/blob/70373384dc117ea387d28a8548d49ecc58e7cca0/api/v1beta1/tags.go#L198-L200

As far as I can see, that tag isn't used anywhere in the codebase for matching or looking up of subnets. It seem purely for display.

Multiple subnets can end up with the same Name value.

**Environment:**

- Cluster-api-provider-aws version: >= 1.5
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by AverageMarcus**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Prevents CAPA from overwriting any user-provided `Name` tags.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3989

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by Ankitasw** [comment]
@AverageMarcus would you be able to continue with the PR?

**Comment by AverageMarcus** [comment]
@Ankitasw I should be able to but don't have the time currently. I'm going to see if any of my colleagues are able to finish it off for me.

**Comment by Ankitasw** [comment]
@AverageMarcus no worries take your time 🙂 just wanted to remind!!!

**Comment by AndiDog** [comment]
@Ankitasw I'm taking over. Added a unit test, and another change that avoids ever having confusion about whether the generated or user-provided subnet `Name` tag is used. The original change in the `Build` function remains because it means that for many other objects (not only subnets), tag building would support a user-provided `Name` tag. I didn't find any code places where the changes are troublesome.

File: api/v1beta1/tags.go [hunk 1]
```diff
@@ -250,6 +250,12 @@ func (b BuildParams) WithCloudProvider(name string) BuildParams {
 // Build builds tags including the cluster tag and returns them in map form.
 func Build(params BuildParams) Tags {
 	tags := make(Tags)
+
```
[the remark this answers is not in the document]
Comment (AndiDog) [reply api/v1beta1/tags.go]: Reverted

**Comment by AndiDog** [comment]
@Ankitasw can you advice on the E2E failures?

**Pull request merged**